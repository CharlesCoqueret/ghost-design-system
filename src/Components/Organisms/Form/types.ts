import { ReactElement } from 'react';

import { IOption } from '../../Atoms/SelectInput';
import { IAmountFieldProps } from '../../Molecules/AmountField/AmountField';
import { IDatePickerFieldProps } from '../../Molecules/DatePickerField/DatePickerField';
import { IFileFieldProps } from '../../Molecules/FileField/FileField';
import { IDynamicSearchFieldProps } from '../../Molecules/DynamicSearchField/DynamicSearchField';
import { IMultiSelectFieldProps } from '../../Molecules/MultiSelectField/MultiSelectField';
import { IPercentageFieldProps } from '../../Molecules/PercentageField/PercentageField';
import { IRichTextFieldProps } from '../../Molecules/RichTextField/RichTextField';
import { ISelectFieldProps } from '../../Molecules/SelectField/SelectField';
import { ILineEditableDataTableProps } from '../DataTable/LineEditableDataTable/LineEditableDataTable';
import { ITextFieldProps } from '../../Molecules/TextField/TextField';
import { ITextAreaFieldProps } from '../../Molecules/TextAreaField/TextAreaField';
import { IYearPickerFieldProps } from '../../Molecules/YearPickerField/YearPickerField';

export interface IFormSubmitReturnedType<T> {
  data: T;
  valid: boolean;
}

export interface IUseFormReturnedType<T> {
  formElement: ReactElement;
  getData: () => T;
  isModified: () => boolean;
  reset: () => void;
  submit: () => IFormSubmitReturnedType<T>;
}

export enum FieldTypeEnum {
  AMOUNT = 'amount',
  CHECKBOX = 'checkbox',
  CUSTOM = 'custom',
  DATE = 'date',
  DESCRIPTION = 'description',
  DYNAMICSEARCH = 'dynamicsearch',
  FILE = 'file',
  MULTISELECT = 'multiselect',
  NUMBER = 'number',
  PERCENTAGE = 'percentage',
  RICHTEXT = 'richtext',
  SECTION = 'section',
  SELECT = 'select',
  SWITCH = 'switch',
  TABLE = 'table',
  TEXT = 'text',
  TEXTAREA = 'textarea',
  YEAR = 'year',
}

export type IFieldAndLayoutProps<T> = IFieldProps<T> | ILayoutProps<T>;

export type ILayoutProps<T> = IFieldSectionProps<T> | IFieldDescriptionProps<T>;

export type IFieldProps<T> =
  | IFieldAmountProps<T>
  | IFieldCheckboxProps<T>
  | IFieldCustomProps<T>
  | IFieldDateProps<T>
  | IFieldDynamicSearchProps<T>
  | IFieldFileProps<T>
  | IFieldMultiSelectProps<T>
  | IFieldNumberProps<T>
  | IFieldPercentageProps<T>
  | IFieldRichtextProps<T>
  | IFieldSelectProps<T>
  | IFieldSwitchProps<T>
  // TODO investigate type resolution
  // Using any to avoid circular type definition for now, until there is a way to get the type of an item of T[keyof T]
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  | IFieldTableProps<T, any>
  | IFieldTextAreaProps<T>
  | IFieldTextProps<T>
  | IFieldYearProps<T>;

export interface IVisibilityProps<T> {
  /** When set to true or when called, returning true, it hides the field or description or section (and children fields) (optional, default: false) */
  hidden?: boolean | ((data: T) => boolean);
}

export interface IFieldBaseProps<T> extends Partial<IVisibilityProps<T>> {
  dataIndex: keyof T;
  disabled?: boolean;
  errorMessage?: string;
  fieldClassName?: string;
  helperText?: string;
  highlighted?: boolean;
  inputClassName?: string;
  label: string;
  mandatory?: boolean;
  name?: string;
  readOnly?: boolean;
}

export interface IFieldAmountProps<T>
  extends IFieldBaseProps<T>,
    Pick<
      IAmountFieldProps,
      | 'allowNegative'
      | 'decimalScale'
      | 'decimalSeparator'
      | 'maxValue'
      | 'minValue'
      | 'placeholder'
      | 'prefix'
      | 'suffix'
      | 'thousandSeparator'
      | 'thousandsGroupStyle'
    > {
  fieldType: FieldTypeEnum.AMOUNT;
}

export interface IFieldCheckboxProps<T> extends IFieldBaseProps<T> {
  fieldType: FieldTypeEnum.CHECKBOX;
}

export interface IFieldCustomProps<T, U = unknown> extends IFieldBaseProps<T> {
  fieldType: FieldTypeEnum.CUSTOM;
  isEqual?: (previousValue: T[keyof T], currentValue: T[keyof T]) => boolean;
  data?: U;
  customField: <
    U extends {
      highlighted?: boolean;
      onChange?: (value: T[keyof T]) => void;
      readOnly?: boolean;
      inputValue?: T[keyof T];
    },
  >(
    props: U,
  ) => ReactElement;
}

export interface IFieldDateProps<T>
  extends IFieldBaseProps<T>,
    Pick<IDatePickerFieldProps, 'calendarStartDay' | 'dateFormat' | 'isClearable' | 'locale' | 'placeholder'> {
  fieldType: FieldTypeEnum.DATE;
}

export interface IFieldDescriptionProps<T> extends Partial<IVisibilityProps<T>> {
  dataIndex?: keyof T;
  description: ReactElement | (<U extends { value: T[keyof T] }>(props: U) => ReactElement);
  fieldType: FieldTypeEnum.DESCRIPTION;
}

export interface IFieldFileProps<T>
  extends IFieldBaseProps<T>,
    Pick<
      IFileFieldProps,
      | 'acceptTypes'
      | 'additionalInfo'
      | 'maxFiles'
      | 'maxFileSize'
      | 'maxFolderDepth'
      | 'onDelete'
      | 'onDownload'
      | 'requestHeaders'
      | 'requestMethod'
      | 'requestUrl'
      | 'requestWithCredentials'
      | 'showFileSize'
      | 'showProgressBar'
      | 'uploadMessage'
      | 'localization'
    > {
  fieldType: FieldTypeEnum.FILE;
}

export interface IFieldDynamicSearchProps<T>
  extends IFieldBaseProps<T>,
    Pick<
      IDynamicSearchFieldProps,
      'colors' | 'isClearable' | 'noOptionsMessage' | 'placeholder' | 'resolveValue' | 'searchOptions'
    > {
  fieldType: FieldTypeEnum.DYNAMICSEARCH;
}

export interface IFieldMultiSelectProps<T>
  extends IFieldBaseProps<T>,
    Pick<
      IMultiSelectFieldProps,
      'colors' | 'isClearable' | 'numberOfItemLabel' | 'numberOfItemsLabel' | 'placeholder'
    > {
  // When the value is not present in the options, should the value be erased (optional, default: false)
  eraseValueWhenNotInOptions?: boolean;
  fieldType: FieldTypeEnum.MULTISELECT;
  options: Array<IOption> | ((data: T) => Array<IOption>);
}

export interface IFieldNumberProps<T>
  extends IFieldBaseProps<T>,
    Pick<
      IAmountFieldProps,
      | 'allowNegative'
      | 'decimalScale'
      | 'decimalSeparator'
      | 'maxValue'
      | 'minValue'
      | 'placeholder'
      | 'prefix'
      | 'suffix'
      | 'thousandSeparator'
      | 'thousandsGroupStyle'
    > {
  fieldType: FieldTypeEnum.NUMBER;
}

export interface IFieldPercentageProps<T>
  extends IFieldBaseProps<T>,
    Pick<
      IPercentageFieldProps,
      | 'allowNegative'
      | 'decimalScale'
      | 'decimalSeparator'
      | 'maxValue'
      | 'minValue'
      | 'placeholder'
      | 'thousandSeparator'
      | 'thousandsGroupStyle'
    > {
  fieldType: FieldTypeEnum.PERCENTAGE;
}

export interface IFieldRichtextProps<T>
  extends IFieldBaseProps<T>,
    Pick<IRichTextFieldProps, 'convertImagesToBase64' | 'enableImage' | 'enableLink' | 'locale' | 'maxLength'> {
  fieldType: FieldTypeEnum.RICHTEXT;
}

export interface IFieldSectionProps<T> extends Partial<IVisibilityProps<T>> {
  collapsable?: boolean;
  fieldType: FieldTypeEnum.SECTION;
  openInitially?: boolean;
  fields: Array<IFieldProps<T>>;
  label: string;
}

export interface IFieldSelectProps<T>
  extends IFieldBaseProps<T>,
    Pick<ISelectFieldProps, 'colors' | 'isClearable' | 'placeholder'> {
  // When the value is not present in the options, should the value be erased (optional, default: false)
  eraseValueWhenNotInOptions?: boolean;
  fieldType: FieldTypeEnum.SELECT;
  options: Array<IOption> | ((data: T) => Array<IOption>);
}

export interface IFieldSwitchProps<T> extends IFieldBaseProps<T> {
  fieldType: FieldTypeEnum.SWITCH;
}

export interface IFieldTableProps<T, U>
  extends IFieldBaseProps<T>,
    Pick<ILineEditableDataTableProps<U>, 'columns' | 'extra' | 'loading' | 'onSortChange'> {
  fieldType: FieldTypeEnum.TABLE;
}

export interface IFieldTextProps<T>
  extends IFieldBaseProps<T>,
    Pick<ITextFieldProps, 'maxLength' | 'minLength' | 'placeholder'> {
  fieldType: FieldTypeEnum.TEXT;
}

export interface IFieldTextAreaProps<T>
  extends IFieldBaseProps<T>,
    Pick<ITextAreaFieldProps, 'maxLength' | 'minLength' | 'placeholder'> {
  fieldType: FieldTypeEnum.TEXTAREA;
}

export interface IFieldYearProps<T> extends IFieldBaseProps<T>, Pick<IYearPickerFieldProps, 'placeholder'> {
  fieldType: FieldTypeEnum.YEAR;
}
