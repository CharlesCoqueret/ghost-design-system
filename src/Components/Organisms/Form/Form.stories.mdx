import { useState } from 'react';
import { Canvas, Meta, ArgsTable, Story } from '@storybook/addon-docs';
import { MemoryRouter as Router } from 'react-router-dom';

import Form from './Form';
import Alert from '../../Molecules/Alert/Alert';

<Meta
  title='Organism/Form'
  component={Form}
  parameters={{
    viewMode: 'docs',
    layout: 'fullscreen',
  }}
/>

# Form

export const Template = (args) => {
  return (
    <Form {...args}>
      <Alert>This is an alert inside a Form</Alert>
    </Form>
  );
};

<Canvas withSource='open'>
  <Story name='Form' args={{}}>
    {Template.bind({})}
  </Story>
</Canvas>

<ArgsTable of={Form} sort='requiredFirst' exclude={['dataTestId']} />
