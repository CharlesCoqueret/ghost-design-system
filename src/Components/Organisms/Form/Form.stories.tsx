import React from 'react';
import * as yup from 'yup';
import cloneDeep from 'lodash/cloneDeep';

import { IToggleEntry } from '../../Atoms/CheckBoxInput/types';
import { Button, ColorButtonEnum } from '../../Molecules';

import useForm, { IUseFormProps } from './useForm';
import { FieldTypeEnum, IFieldAndLayoutProps } from './types';
import { Link, Typography } from '../../Atoms';

export default {
  title: 'Organism/Form',
};

interface dataType {
  amount: number | undefined;
  checkbox: Array<IToggleEntry>;
  date: Date | undefined | null;
  description?: string;
  multiselect: Array<string> | undefined;
  number: number | undefined;
  percentage: number | undefined;
  select: string | undefined;
  switch: Array<IToggleEntry> | undefined;
  text: string | undefined;
  textarea: string | undefined;
  year: number | undefined;
}

const checkboxOption = [
  { value: 'value 1', label: 'label 1' },
  { value: 'value 2', checked: true, label: 'label 2' },
];

const switchOption = [
  { value: 'value 1', label: 'label 1' },
  { value: 'value 2', checked: true, label: 'label 2' },
];

const options = [
  { value: 'CASE', label: 'case' },
  { value: 'CLAIM', label: 'claim' },
  { value: 'COMPANY', label: 'company' },
  { value: 'COUNTRY', label: 'country' },
];

const initialData: dataType = {
  amount: 100000,
  checkbox: cloneDeep(checkboxOption),
  date: new Date(),
  description: 'Description',
  multiselect: options.map((option) => option.value),
  number: 10,
  percentage: 50,
  select: options[2].value,
  switch: cloneDeep(switchOption),
  text: 'text',
  textarea:
    'Lorem ipsum dolor sit amet. Ut voluptas reiciendis vel praesentium laborum hic voluptas asperiores nam ' +
    'rerum nihil obcaecati labore. Id praesentium porro ea placeat rerum aut tempore totam aut illum cupiditate sed ' +
    'laborum explicabo. Hic explicabo voluptatibus qui repellat fugiat ex voluptatum fuga qui architecto atque quo ' +
    'illum quas aut facilis nesciunt? Ut suscipit rerum ut perferendis nihil ea autem unde est enim veniam nam odio ' +
    'tempora.\n\nQuo numquam iste est repellendus numquam et galisum omnis ad praesentium dolores aut neque saepe ' +
    'vel consectetur enim aut cumque neque. Et voluptate sapiente quisquam quasi eum beatae voluptas rem iure velit. ' +
    'Sed impedit eaque 33 natus nihil est quaerat porro est quia nisi qui doloribus aperiam. Sit culpa illum ea ' +
    'consectetur perspiciatis ex veritatis dolorem id velit sequi qui maiores asperiores!',
  year: 1984,
};

const oldData: dataType = {
  amount: 10000,
  checkbox: cloneDeep(checkboxOption).map((option) => {
    option.checked = !option.checked;
    return option;
  }),
  date: new Date('09/24/1984'),
  multiselect: options.filter((_option, index) => index % 2 === 0).map((option) => option.value),
  number: 11,
  percentage: 51,
  select: options[3].value,
  switch: cloneDeep(switchOption).map((option) => {
    option.checked = !option.checked;
    return option;
  }),
  text: 'old text',
  textarea:
    'Lorem ipsum dolor sit amet. Ut voluptas reiciendis vel praesentium laborum hic voluptas asperiores nam ' +
    'rerum nihil obcaecati labore. Id praesentium porro ea placeat rerum aut tempore totam aut illum cupiditate sed ' +
    'laborum explicabo. Hic explicabo voluptatibus qui repellat fugiat ex voluptatum fuga qui architecto atque quo ' +
    'illum quas aut facilis nesciunt? Ut suscipit rerum ut perferendis nihil ea autem unde est enim veniam nam odio ' +
    'tempora.\n\nQuo numquam iste est repellendus numquam et galisum omnis ad praesentium dolores aut neque saepe ' +
    'vel consectetur enim aut cumque neque. Et voluptate sapiente quisquam quasi eum beatae voluptas rem iure velit. ' +
    'Sed impedit eaque 33 natus nihil est quaerat porro est quia nisi qui doloribus aperiam. Sit culpa illum ea ' +
    'consectetur perspiciatis ex veritatis dolorem id velit sequi qui maiores asperiores!\n\n' +
    'Lorem ipsum dolor sit amet. Ut voluptas reiciendis vel praesentium laborum hic voluptas asperiores nam ' +
    'rerum nihil obcaecati labore. Id praesentium porro ea placeat rerum aut tempore totam aut illum cupiditate sed ' +
    'laborum explicabo. Hic explicabo voluptatibus qui repellat fugiat ex voluptatum fuga qui architecto atque quo ' +
    'illum quas aut facilis nesciunt? Ut suscipit rerum ut perferendis nihil ea autem unde est enim veniam nam odio ' +
    'tempora.\n\nQuo numquam iste est repellendus numquam et galisum omnis ad praesentium dolores aut neque saepe ' +
    'vel consectetur enim aut cumque neque. Et voluptate sapiente quisquam quasi eum beatae voluptas rem iure velit. ' +
    'Sed impedit eaque 33 natus nihil est quaerat porro est quia nisi qui doloribus aperiam. Sit culpa illum ea ' +
    'consectetur perspiciatis ex veritatis dolorem id velit sequi qui maiores asperiores!',
  year: 1994,
};

const validationSchema = yup.object({
  amount: yup.number().required(),
  checkbox: yup
    .array()
    .of(
      yup.object().shape({
        checked: yup.boolean(),
        value: yup.string(),
        label: yup.string(),
        highlighted: yup.boolean().optional(),
      }),
    )
    .test({
      name: 'one-true',
      message: 'At least one required',
      test: (val) => val.some((entry) => entry.checked === true),
    })
    .required(),
  date: yup.date().min(new Date('01/01/1980'), 'Date needs to be after Jan 1 1980').required(),
  description: yup.string().optional(),
  multiselect: yup.array().of(yup.string()).min(1, 'At least one item required').required(),
  number: yup.number().required('Value for number is required').min(5, 'Minimum value is 5'),
  percentage: yup.number().required(),
  select: yup.string().required(),
  switch: yup
    .array()
    .of(
      yup.object().shape({
        checked: yup.boolean(),
        value: yup.string(),
        label: yup.string(),
        highlighted: yup.boolean().optional(),
      }),
    )
    .test({
      name: 'one-true',
      message: 'At least one required',
      test: (val) => val.some((entry) => entry.checked === true),
    })
    .required(),
  text: yup
    .string()
    .required()
    .transform((value) => value.trim())
    .min(3, 'minimun length is 3 without spaces at the end of beginning'),
  textarea: yup
    .string()
    .required()
    .transform((value) => value.trim())
    .min(3, 'minimun length is 10 without spaces at the end of beginning'),
  year: yup.number().min(1984, 'Date needs to be after 1984').required(),
});

const fields: Array<IFieldAndLayoutProps<dataType>> = [
  {
    description: (
      <div>
        <Typography.Text>Any description</Typography.Text>
        <Link link='http://google.com' text='external link' />
        <Link link='#' text='internal link' />
      </div>
    ),
    fieldType: FieldTypeEnum.DESCRIPTION,
  },
  {
    label: 'Amount',
    dataIndex: 'amount',
    fieldType: FieldTypeEnum.AMOUNT,
  },
  { label: 'Checkbox', dataIndex: 'checkbox', fieldType: FieldTypeEnum.CHECKBOX },
  { label: 'Date', dataIndex: 'date', fieldType: FieldTypeEnum.DATE },
  { label: 'Multiselect', dataIndex: 'multiselect', fieldType: FieldTypeEnum.MULTISELECT, options: options },
  {
    label: 'Section',
    fieldType: FieldTypeEnum.SECTION,
    openInitially: true,
    fields: [
      {
        label: 'Number',
        dataIndex: 'number',
        fieldType: FieldTypeEnum.NUMBER,
        mandatory: true,
      },
      {
        label: 'Percentage',
        dataIndex: 'percentage',
        fieldType: FieldTypeEnum.PERCENTAGE,
      },
    ],
  },
  {
    label: 'Additional Section',
    fieldType: FieldTypeEnum.SECTION,
    openInitially: true,
    fields: [
      { label: 'Select', dataIndex: 'select', fieldType: FieldTypeEnum.SELECT, options: options },
      { label: 'Switch', dataIndex: 'switch', fieldType: FieldTypeEnum.SWITCH },
      { label: 'Text', dataIndex: 'text', fieldType: FieldTypeEnum.TEXT },
      { label: 'Textarea', dataIndex: 'textarea', fieldType: FieldTypeEnum.TEXTAREA },
      { label: 'Year', dataIndex: 'year', fieldType: FieldTypeEnum.YEAR },
    ],
  },
];

const Template = (args: IUseFormProps<dataType>) => {
  const { formElement, submit, reset } = useForm<dataType>({
    initialData: initialData,
    fields: fields,
    validationSchema: validationSchema,
  });
  return (
    <>
      <div style={{ display: 'flex', justifyContent: 'space-around' }}>
        <Button
          label='Submit'
          onClick={() => {
            console.log('submit', JSON.stringify(submit()));
          }}
          color={ColorButtonEnum.PRIMARY}
        />
        <Button
          label='Reset'
          onClick={() => {
            console.log('reset', JSON.stringify(reset()));
          }}
          color={ColorButtonEnum.SECONDARY}
        />
      </div>
      <div>{formElement}</div>
    </>
  );
};

const HighlightedTemplate = (args: IUseFormProps<dataType>) => {
  const { formElement, getData, isModified, submit, reset } = useForm<dataType>({
    initialData: initialData,
    fields: fields,
    validationSchema: validationSchema,
    previousData: oldData,
  });

  return (
    <>
      <div style={{ display: 'flex', justifyContent: 'space-around' }}>
        <Button
          label='Submit'
          onClick={() => {
            console.log('submit', JSON.stringify(submit()));
          }}
          color={ColorButtonEnum.PRIMARY}
        />
        <Button
          label='Reset'
          onClick={() => {
            console.log('reset', JSON.stringify(reset()));
          }}
          color={ColorButtonEnum.SECONDARY}
        />
      </div>
      <div>{formElement}</div>
      <div>
        <pre>Has been modified: {isModified().toString()}</pre>
      </div>
      <div>
        Current data:
        <textarea style={{ width: '100%', height: '300px' }} value={JSON.stringify(getData(), null, 2)} readOnly />
      </div>
    </>
  );
};

export const Default = Template.bind({});
Default.args = {};

export const Highlighted = HighlightedTemplate.bind({});
Highlighted.args = {};
