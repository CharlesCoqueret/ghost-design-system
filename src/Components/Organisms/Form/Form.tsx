import React, { Fragment, memo, ReactElement } from 'react';
import * as yup from 'yup';
import { AnyObject } from 'yup/lib/object';
import { SchemaDescription, SchemaObjectDescription } from 'yup/lib/schema';

import { FieldTypeEnum, IFieldAndLayoutProps } from './types';
import FormField from './FormField';
import { FieldError } from './yupResolver';
import { Col, Row, Section } from '../../Atoms/Layout';

export interface IFormProps<T extends AnyObject> {
  enableOldData?: boolean;
  enableSideBySide?: boolean;
  fields: Array<IFieldAndLayoutProps<T>>;
  handleDataChange: (dataIndex: keyof T, newValue: T[keyof T]) => void;
  initialData: T;
  previousData?: T;
  validationError?: Partial<Record<keyof T, FieldError>>;
  validationSchema?: yup.SchemaOf<T>;
  usePortal?: boolean;
}

const Form = <T extends AnyObject>(props: IFormProps<T>): ReactElement => {
  const {
    enableOldData,
    enableSideBySide,
    fields,
    handleDataChange,
    initialData,
    previousData,
    validationSchema,
    validationError,
    usePortal,
  } = props;

  const localUsePortal = usePortal === undefined ? true : usePortal;

  return (
    <Col>
      {fields.map((field, index) => {
        // Managing hidden field
        if (
          field.hidden !== undefined &&
          ((typeof field.hidden === 'function' && field.hidden(initialData) === true) || field.hidden === true)
        ) {
          return <Fragment key={`hidden-${index}`}></Fragment>;
        }

        if (field.fieldType === FieldTypeEnum.SECTION) {
          return (
            <Section
              key={`section-${field.label}`}
              level={field.level}
              title={field.label}
              collapsible={field.collapsible}
              openInitially={field.openInitially}
              separator={field.separator}>
              <Form
                enableOldData={enableOldData}
                enableSideBySide={enableSideBySide}
                fields={field.fields}
                handleDataChange={handleDataChange}
                initialData={initialData}
                previousData={previousData}
                validationError={validationError}
                validationSchema={validationSchema}
                usePortal={localUsePortal}
              />
            </Section>
          );
        }
        if (field.fieldType === FieldTypeEnum.DESCRIPTION) {
          return <Row key={`description-${index}`}>{field.description}</Row>;
        }

        let isRequired = false;

        try {
          const objectDescription = validationSchema?.describe() as SchemaObjectDescription | undefined;
          const schemaDescription = objectDescription?.fields[field.dataIndex as string] as SchemaDescription;
          isRequired = schemaDescription.tests.some((test) => test.name === 'required');
        } catch {
          console.warn(`could not retrieve if ${JSON.stringify(field.dataIndex)} is mandatory`);
        }

        return (
          <FormField<T>
            enableOldData={enableOldData}
            enableSideBySide={enableSideBySide}
            key={`field-${field.label}`}
            field={field}
            data={initialData}
            previousData={previousData}
            handleChange={handleDataChange}
            validationError={validationError}
            requiredFromValidation={isRequired}
            usePortal={localUsePortal}
          />
        );
      })}
    </Col>
  );
};

export default memo(Form) as typeof Form;
