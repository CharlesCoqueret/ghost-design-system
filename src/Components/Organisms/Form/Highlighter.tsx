import React, { memo, PropsWithChildren, ReactElement } from 'react';

import { Col, Row } from '../../Atoms/Layout';

export interface IHighlighterProps {
  enableOldData?: boolean;
  enableSideBySide?: boolean;
  shouldHighlight?: boolean;
  oldData?: unknown;
}

const Highlighter = (props: PropsWithChildren<IHighlighterProps>): ReactElement => {
  const { children, enableOldData, enableSideBySide, oldData, shouldHighlight } = props;

  if (!enableSideBySide) return <>{children}</>;

  return (
    <>
      {React.Children.map(children, (child) => {
        if (!React.isValidElement(child)) return child;
        return (
          <Row>
            <Col>{child}</Col>
            <Col>
              {enableOldData &&
                React.cloneElement(child, {
                  disabled: undefined,
                  errorMessage: undefined,
                  extra: {
                    ...child.props?.extra,
                    // For table use case
                    onEdit: undefined,
                    onRowEdit: undefined,
                    onRowSubmit: undefined,
                    onRowDelete: undefined,
                    isEditable: () => false,
                    isDeletable: () => false,
                    canAddNewLine: () => false,
                  },
                  helperText: undefined,
                  highlighted: shouldHighlight,
                  id: undefined,
                  inputValue: oldData, // For fields
                  data: oldData || [], // For table use case
                  maxLength: undefined,
                  name: undefined,
                  onChange: undefined,
                  placeholder: undefined,
                  readOnly: true,
                })}
            </Col>
          </Row>
        );
      })}
    </>
  );
};

export default memo(Highlighter);
