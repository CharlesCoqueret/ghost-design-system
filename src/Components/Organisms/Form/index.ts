export type { IFieldProps, IFormSubmitReturnedType, IUseFormReturnedType } from './types';
export { FieldTypeEnum } from './types';
export { default as Form } from './Form';
export { default as Highlighter } from './Highlighter';
export { default as useForm } from './useForm';
