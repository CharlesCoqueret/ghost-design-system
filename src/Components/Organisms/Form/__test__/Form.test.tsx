import React from 'react';
import { render } from '@testing-library/react';
import * as yup from 'yup';

// Mocking suneditor which is problematic with Jest
jest.mock('suneditor', () => {});
jest.mock('suneditor/src/plugins/', () => {});
jest.mock('suneditor/src/plugins/submenu/align', () => {});
jest.mock('suneditor/src/plugins/command/blockquote', () => {});
jest.mock('suneditor/src/plugins/submenu/fontColor', () => {});
jest.mock('suneditor/src/plugins/submenu/fontSize', () => {});
jest.mock('suneditor/src/plugins/submenu/formatBlock', () => {});
jest.mock('suneditor/src/plugins/submenu/hiliteColor', () => {});
jest.mock('suneditor/src/plugins/submenu/horizontalRule', () => {});
jest.mock('suneditor/src/plugins/dialog/image', () => {});
jest.mock('suneditor/src/plugins/dialog/link', () => {});
jest.mock('suneditor/src/plugins/submenu/lineHeight', () => {});
jest.mock('suneditor/src/plugins/submenu/list', () => {});
jest.mock('suneditor/src/plugins/submenu/paragraphStyle', () => {});
jest.mock('suneditor/src/plugins/submenu/table', () => {});
jest.mock('suneditor-react', () => {});
jest.mock('suneditor-react/dist', () => {});
jest.mock('suneditor-react/dist/types/lang', () => {});

import Form from '../Form';
import { FieldTypeEnum } from '../types';

describe('Form Component', () => {
  it('Form renders with description and sections', () => {
    const handleDataChangeMock = jest.fn();
    const { container } = render(
      <Form
        fields={[
          {
            label: 'Title',
            fieldType: FieldTypeEnum.SECTION,
            fields: [],
          },
          {
            description: <>Description</>,
            fieldType: FieldTypeEnum.DESCRIPTION,
          },
          {
            description: <>Hidden function description</>,
            fieldType: FieldTypeEnum.DESCRIPTION,
            hidden: () => true,
          },
          {
            description: <>Hidden description</>,
            fieldType: FieldTypeEnum.DESCRIPTION,
            hidden: true,
          },
        ]}
        handleDataChange={handleDataChangeMock}
        initialData={{}}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('Form renders with mandatory field', () => {
    const handleDataChangeMock = jest.fn();
    const { container } = render(
      <Form<{ number?: number }>
        fields={[
          {
            label: 'Number',
            fieldType: FieldTypeEnum.NUMBER,
            dataIndex: 'number',
          },
        ]}
        handleDataChange={handleDataChangeMock}
        initialData={{ number: undefined }}
        validationSchema={yup.object({ number: yup.number().required('Value for number is required') })}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('Form renders without validation schema', () => {
    const handleDataChangeMock = jest.fn();
    const { container } = render(
      <Form<{ number?: number }>
        fields={[
          {
            label: 'Number',
            fieldType: FieldTypeEnum.NUMBER,
            dataIndex: 'number',
          },
        ]}
        handleDataChange={handleDataChangeMock}
        initialData={{ number: undefined }}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
