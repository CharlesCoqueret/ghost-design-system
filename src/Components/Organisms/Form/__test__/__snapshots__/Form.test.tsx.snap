// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component renders properly 1`] = `
<div>
  <form
    class="form CLASSNAME"
  >
    <div>
      CHILD-1
    </div>
  </form>
</div>
`;
